### ISSUE TITLE
get_zone_states on LXIVM controller

### ISSUE THREAD

**Issue by funkzy** [allenporter/pyrainbird#315]
Hi,

I have the LXIVM two wire controller and would like to write my own interface for controlling the device. I noticed that out of the box, some of the pyrainbird exposed API's work and some do not. (For reference, the LXIVM uses the LNK2 wifi module.)

For get_zone_states, stations 1-24 work:

(running zone 9)
$ python3 rainbird_tool.py get_zone_states
states: 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, **9:1**, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0, 16:0, 17:0, 18:0, 19:0, 20:0, 21:0, 22:0, 23:0, 24:0

But running zone 27 is not displayed:

$ python3 rainbird_tool.py get_zone_states
states: 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0, 16:0, 17:0, 18:0, 19:0, 20:0, 21:0, 22:0, 23:0, 24:0

Any suggestions on how to display > 24 stations?

Thanks!

**Comment by allenporter** [comment]
Thanks for the report.

Since i don't have one of these devices, we'll want to capture some of the data here.  Maybe running with `--log-level=debug` will give us enough detail to understand the change needed?

**Comment by allenporter** [comment]
Otherwise you can see https://github.com/allenporter/pyrainbird/blob/main/CONTRIBUTING.md for how to start up a proxy to get request traces from the app. See discussion in https://github.com/allenporter/pyrainbird/issues/178

**Comment by funkzy** [comment]
Awesome - and thanks for developing this awesome tool!

### Debug log for no zones running:

$ python3 rainbird_tool.py --log-level=debug get_zone_states
DEBUG:pyrainbird.async_client:Request (CurrentStationsActiveRequest): 3F00
DEBUG:pyrainbird.async_client:Request: {"id": 1702318161.536479, "jsonrpc": "2.0", "method": "tunnelSip", "params": {"data": "3F00", "length": 2}}
DEBUG:pyrainbird.async_client:Response: {"id":1702318161.536479,"jsonrpc":"2.0","result":{"data":"BF0000000000","length":6}}
DEBUG:pyrainbird.async_client:Response from line: BF0000000000
DEBUG:pyrainbird.async_client:Response: {'type': 'CurrentStationsActiveResponse', 'pageNumber': 0, 'activeStations': 0}
states: 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0, 16:0, 17:0, 18:0, 19:0, 20:0, 21:0, 22:0, 23:0, 24:0

### Debug log w/ zone 9 running:
$ python3 rainbird_tool.py --log-level=debug get_zone_states
DEBUG:pyrainbird.async_client:Request (CurrentStationsActiveRequest): 3F00
DEBUG:pyrainbird.async_client:Request: {"id": 1702318650.676683, "jsonrpc": "2.0", "method": "tunnelSip", "params": {"data": "3F00", "length": 2}}
DEBUG:pyrainbird.async_client:Response: {"id":1702318650.676683,"jsonrpc":"2.0","result":{"data":"BF0000010000","length":6}}
DEBUG:pyrainbird.async_client:Response from line: BF0000010000
DEBUG:pyrainbird.async_client:Response: {'type': 'CurrentStationsActiveResponse', 'pageNumber': 0, 'activeStations': 65536}
states: 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:1, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0, 16:0, 17:0, 18:0, 19:0, 20:0, 21:0, 22:0, 23:0, 24:0

### Debug log w/ zone 25 running:
$ python3 rainbird_tool.py --log-level=debug get_zone_states
DEBUG:pyrainbird.async_client:Request (CurrentStationsActiveRequest): 3F00
DEBUG:pyrainbird.async_client:Request: {"id": 1702318480.244543, "jsonrpc": "2.0", "method": "tunnelSip", "params": {"data": "3F00", "length": 2}}
DEBUG:pyrainbird.async_client:Response: {"id":1702318480.244543,"jsonrpc":"2.0","result":{"data":"BF0000000001","length":6}}
DEBUG:pyrainbird.async_client:Response from line: BF0000000001
DEBUG:pyrainbird.async_client:Response: {'type': 'CurrentStationsActiveResponse', 'pageNumber': 0, 'activeStations': 1}
states: 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0, 16:0, 17:0, 18:0, 19:0, 20:0, 21:0, 22:0, 23:0, 24:0

### Debug log w/ zone 26 running:
$ python3 rainbird_tool.py --log-level=debug get_zone_states
DEBUG:pyrainbird.async_client:Request (CurrentStationsActiveRequest): 3F00
DEBUG:pyrainbird.async_client:Request: {"id": 1702318976.519146, "jsonrpc": "2.0", "method": "tunnelSip", "params": {"data": "3F00", "length": 2}}
DEBUG:pyrainbird.async_client:Response: {"id":1702318976.519146,"jsonrpc":"2.0","result":{"data":"BF0000000002","length":6}}
DEBUG:pyrainbird.async_client:Response from line: BF0000000002
DEBUG:pyrainbird.async_client:Response: {'type': 'CurrentStationsActiveResponse', 'pageNumber': 0, 'activeStations': 2}
states: 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0, 16:0, 17:0, 18:0, 19:0, 20:0, 21:0, 22:0, 23:0, 24:0

### Debug log w/ zone 27 running:
$ python3 rainbird_tool.py --log-level=debug get_zone_states
DEBUG:pyrainbird.async_client:Request (CurrentStationsActiveRequest): 3F00
DEBUG:pyrainbird.async_client:Request: {"id": 1702318194.8170502, "jsonrpc": "2.0", "method": "tunnelSip", "params": {"data": "3F00", "length": 2}}
DEBUG:pyrainbird.async_client:Response: {"id":1702318194.8170502,"jsonrpc":"2.0","result":{"data":"BF0000000004","length":6}}
DEBUG:pyrainbird.async_client:Response from line: BF0000000004
DEBUG:pyrainbird.async_client:Response: {'type': 'CurrentStationsActiveResponse', 'pageNumber': 0, 'activeStations': 4}
states: 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0, 12:0, 13:0, 14:0, 15:0, 16:0, 17:0, 18:0, 19:0, 20:0, 21:0, 22:0, 23:0, 24:0


### Debug log w/ zones 12, 13, 14, and 21 running in parallel:

The controller also allows zones to run in parallel, so here are four zones (12, 13, 14, and 21) running in parallel:

$ python3 rainbird_tool.py --log-level=debug get_zone_states
DEBUG:pyrainbird.async_client:Request (CurrentStationsActiveRequest): 3F00
DEBUG:pyrainbird.async_client:Request: {"id": 1702319091.173444, "jsonrpc": "2.0", "method": "tunnelSip", "params": {"data": "3F00", "length": 2}}
DEBUG:pyrainbird.async_client:Response: {"id":1702319091.173444,"jsonrpc":"2.0","result":{"data":"BF0000381000","length":6}}
DEBUG:pyrainbird.async_client:Response from line: BF0000381000
DEBUG:pyrainbird.async_client:Response: {'type': 'CurrentStationsActiveResponse', 'pageNumber': 0, 'activeStations': 3674112}
states: 1:0, 2:0, 3:0, 4:0, 5:0, 6:0, 7:0, 8:0, 9:0, 10:0, 11:0, 12:1, 13:1, 14:1, 15:0, 16:0, 17:0, 18:0, 19:0, 20:0, 21:1, 22:0, 23:0, 24:0

**Comment by funkzy** [comment]
Just getting back to this now - the current state of the union is:

- get_zone_states displays the status of stations > 24 (woohoo!)
- I can't irrigate_zone for either zone 23 or zone 28 and strangely they produce different errors:

### Irrigate zone 23:

$ python3 rainbird_tool.py --log-level=debug irrigate_zone 23 1
Traceback (most recent call last):
  File "/Users/x/python_projects/quickbird/rainbird_tool.py", line 109, in <module>
    asyncio.run(main())
    ~~~~~~~~~~~^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ~~~~~~~~~~^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/asyncio/base_events.py", line 720, in run_until_complete
    return future.result()
           ~~~~~~~~~~~~~^^
  File "/Users/x/python_projects/quickbird/rainbird_tool.py", line 103, in main
    result = await args.func(client, **method_args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/x/python_projects/quickbird/lib/python3.13/site-packages/pyrainbird/async_client.py", line 327, in irrigate_zone
    await self._process_command(
        lambda resp: True, "ManuallyRunStationRequest", zone, minutes
    )
  File "/Users/x/python_projects/quickbird/lib/python3.13/site-packages/pyrainbird/async_client.py", line 481, in _process_command
    data = rainbird.encode(command, *args)
  File "/Users/x/python_projects/quickbird/lib/python3.13/site-packages/pyrainbird/rainbird.py", line 196, in encode
    return encode_command(command_set, *args)
  File "/Users/x/python_projects/quickbird/lib/python3.13/site-packages/pyrainbird/rainbird.py", line 213, in encode_command
    params = (cmd_code,) + tuple(map(lambda x: int(x), args))
                           ~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/x/python_projects/quickbird/lib/python3.13/site-packages/pyrainbird/rainbird.py", line 213, in <lambda>
    params = (cmd_code,) + tuple(map(lambda x: int(x), args))
                                               ~~~^^^
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'datetime.time'


### Irrigate zone 28:

$ python3 rainbird_tool.py --log-level=debug irrigate_zone 28 1
DEBUG:pyrainbird.async_client:Request (ManuallyRunStationRequest): 39002801
DEBUG:pyrainbird.async_client:Request: {"id": 1739750821.3688939, "jsonrpc": "2.0", "method": "tunnelSip", "params": {"data": "39002801", "length": 4}}
DEBUG:pyrainbird.async_client:Response: '{"id":1739750821.3688939,"jsonrpc":"2.0","result":{"data":"003904","length":3}}'
DEBUG:pyrainbird.async_client:Response from line: 003904
DEBUG:pyrainbird.async_client:Response: {'type': 'NotAcknowledgeResponse', 'commandEcho': 57, 'NAKCode': 4}
ERROR:pyrainbird.async_client:Request (ManuallyRunStationRequest) failed with wrong response! Requested ({'01'}), got 00:
{'type': 'NotAcknowledgeResponse', 'commandEcho': 57, 'NAKCode': 4}
Traceback (most recent call last):
  File "/Users/x/python_projects/quickbird/rainbird_tool.py", line 109, in <module>
    asyncio.run(main())
    ~~~~~~~~~~~^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ~~~~~~~~~~^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/asyncio/base_events.py", line 720, in run_until_complete
    return future.result()
           ~~~~~~~~~~~~~^^
  File "/Users/x/python_projects/quickbird/rainbird_tool.py", line 103, in main
    result = await args.func(client, **method_args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/x/python_projects/quickbird/lib/python3.13/site-packages/pyrainbird/async_client.py", line 327, in irrigate_zone
    await self._process_command(
        lambda resp: True, "ManuallyRunStationRequest", zone, minutes
    )
  File "/Users/x/python_projects/quickbird/lib/python3.13/site-packages/pyrainbird/async_client.py", line 500, in _process_command
    raise RainbirdApiException("Unexpected response from Rain Bird device")
pyrainbird.exceptions.RainbirdApiException: Unexpected response from Rain Bird device





Any ideas on what might be going on? Thank you!

**Comment by allenporter** [comment]
I think this is having a problem because the client library is sending the request incorrectly. I think we need to capture this happening from a real device with mitm proxy to understand how the requests and responses should look since they client is not supporting it right yet.

**Comment by funkzy** [comment]
Sounds good - I'll get a capture going.

**Comment by funkzy** [comment]
Sent via email

**Comment by funkzy** [comment]
Here is the payload sent to run zone 22:


--- Decrypted ---
{"id":1,"jsonrpc":"2.0","method":"tunnelSip","params":{"data":"4B001603","length":4}}
--- request params ---
{'data': '4B001603', 'length': 4}
--- decoded request data ---
{'type': 'StackManuallyRunStationRequest', 'page': 0, 'zone': 22, 'minutes': 3}


And here is the payload sent to run zone 28:

--- Decrypted ---
{"id":1,"jsonrpc":"2.0","method":"tunnelSip","params":{"data":"4B001C03","length":4}}
--- request params ---
{'data': '4B001C03', 'length': 4}
--- decoded request data ---
{'type': 'StackManuallyRunStationRequest', 'page': 0, 'zone': 28, 'minutes': 3}

So it appears that the '16' and '1C' byte is a simple hex encoding of the zone number?

**Comment by allenporter** [comment]
Thanks, looks like i can't view the raw dump you sent since the packets are encoded with the device password.  Maybe you can send again the decoded irrigation request if you can find it?

**Comment by funkzy** [comment]
Yes, is there a way to decode and save the entire capture?

**Comment by allenporter** [comment]
I don't know but there are some videos on the mitm site now https://docs.mitmproxy.org/stable/web-tutorials/web-01-user-interface/

**Comment by allenporter** [comment]
We're moving to support the cloud APIs now, so this should be handled there automatically. please reopen a new issue if you want to discuss anything further that wasn't addressed here.
